packages/edit-site/src/components/add-new-template/add-custom-template-modal.js

```diff
@@ -0,0 +1,243 @@
+/**
+ * WordPress dependencies
+ */
+import { useState, useMemo, useEffect } from '@wordpress/element';
+import { __, sprintf } from '@wordpress/i18n';
+import {
+	Button,
+	Flex,
+	FlexItem,
+	Modal,
+	SearchControl,
+	TextHighlight,
+	__experimentalText as Text,
+	__experimentalHeading as Heading,
+	__unstableComposite as Composite,
+	__unstableUseCompositeState as useCompositeState,
+	__unstableCompositeItem as CompositeItem,
+} from '@wordpress/components';
+import { useSelect } from '@wordpress/data';
+import { useDebounce } from '@wordpress/compose';
+import { store as coreStore } from '@wordpress/core-data';
+
+/**
+ * Internal dependencies
+ */
+import { mapToIHasNameAndId } from './utils';
+
+const EMPTY_ARRAY = [];
+const BASE_QUERY = {
+	order: 'asc',
+	_fields: 'id,title,slug,link',
+	context: 'view',
+};
+
+function SuggestionListItem( {
+	suggestion,
+	search,
+	onSelect,
+	entityForSuggestions,
+	composite,
+} ) {
+	const baseCssClass =
+		'edit-site-custom-template-modal__suggestions_list__list-item';
+	return (
+		<CompositeItem
+			role="option"
+			as={ Button }
+			{ ...composite }
+			className={ baseCssClass }
+			onClick={ () => {
+				const title = `${ entityForSuggestions.labels.singular }: ${ suggestion.name }`;
+				onSelect( {
+					title,
+					description: `Template for ${ title }`,
+					slug: `single-${ entityForSuggestions.slug }-${ suggestion.slug }`,
+				} );
+			} }
+		>
+			<span className={ `${ baseCssClass }__title` }>
+				<TextHighlight text={ suggestion.name } highlight={ search } />
+			</span>
+			{ suggestion.link && (
+				<span className={ `${ baseCssClass }__info` }>
+					{ suggestion.link }
+				</span>
+			) }
+		</CompositeItem>
+	);
+}
+
+function SuggestionList( { entityForSuggestions, onSelect } ) {
+	const composite = useCompositeState( { orientation: 'vertical' } );
+	const [ suggestions, setSuggestions ] = useState( EMPTY_ARRAY );
+	// We need to track two values, the search input's value(searchInputValue)
+	// and the one we want to debounce(search) and make REST API requests.
+	const [ searchInputValue, setSearchInputValue ] = useState( '' );
+	const [ search, setSearch ] = useState( '' );
+	const debouncedSearch = useDebounce( setSearch, 250 );
+	const { searchResults, searchHasResolved } = useSelect(
+		( select ) => {
+			const { getEntityRecords, hasFinishedResolution } =
+				select( coreStore );
+			const selectorArgs = [
+				entityForSuggestions.type,
+				entityForSuggestions.slug,
+				{
+					...BASE_QUERY,
+					search,
+					orderby: !! search ? 'relevance' : 'modified',
+					exclude: entityForSuggestions.postsToExclude,
+					per_page: !! search ? 20 : 10,
+				},
+			];
+			return {
+				searchResults: getEntityRecords( ...selectorArgs ),
+				searchHasResolved: hasFinishedResolution(
+					'getEntityRecords',
+					selectorArgs
+				),
+			};
+		},
+		[ search ]
+	);
```

Comment:
I've tested and you should be able to use `useEntityRecords`:

```js
const { records, hasResolved } = useEntityRecords(
	entityForSuggestions.type,
	entityForSuggestions.slug,
	{
		...BASE_QUERY,
		search,
		orderby: !! search ? 'relevance' : 'modified',
		exclude: entityForSuggestions.postsToExclude,
		per_page: !! search ? 20 : 10,
	}
)
```

Just make sure it correctly reacts to changes to `search` and that it doesn't re-query too often. In particular, see if `queryArgs` (third argument) needs to be explicitly memoized by us or not.

Reply:
Good suggestion! I had forgotten about this new hook and after testing, it looks good here. Also I don't think we need to do anything more about the `search`,  as the query is serialized and passed as a dependency in the hook